File: solver.py
#!/usr/bin/env python3
"""Exact Sudoku precoloring extension on its constraint graph. Standard library only."""

from __future__ import annotations

import argparse
import hashlib
import json
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parent
DEMO = "530070000600195000098000060800060003400803001700020006060000280000419005000080079"
COLORS = frozenset(range(1, 10))


class PuzzleError(ValueError):
    pass


def parse_puzzle(value):
    if isinstance(value, str):
        value = "".join(c for c in value if not c.isspace() and c not in "|+-")
        if len(value) != 81 or any(c not in "0123456789." for c in value):
            raise PuzzleError("Enter exactly 81 digits; use 0 or . for an empty cell.")
        return [0 if c == "." else int(c) for c in value]
    if not isinstance(value, (list, tuple)) or len(value) != 81:
        raise PuzzleError("A puzzle must contain 81 cells.")
    if any(type(v) is not int or not 0 <= v <= 9 for v in value):
        raise PuzzleError("Each cell must be an integer from 0 to 9.")
    return list(value)


def box(i):
    return i // 27 * 3 + i % 9 // 3


def build_graph():
    """One vertex per cell; edges encode row, column, or box inequality."""
    neighbors = [set() for _ in range(81)]
    for i in range(81):
        for j in range(i + 1, 81):
            if i // 9 == j // 9 or i % 9 == j % 9 or box(i) == box(j):
                neighbors[i].add(j)
                neighbors[j].add(i)
    edges = [(i, j) for i, ns in enumerate(neighbors) for j in sorted(ns) if i < j]
    return neighbors, edges


NEIGHBORS, EDGES = build_graph()


def validate_givens(puzzle):
    for i, j in EDGES:
        if puzzle[i] and puzzle[i] == puzzle[j]:
            raise PuzzleError(
                f"Conflicting givens at r{i // 9 + 1}c{i % 9 + 1} and "
                f"r{j // 9 + 1}c{j % 9 + 1}: both are {puzzle[i]}."
            )


def legal_colors(state, vertex):
    return sorted(COLORS - {state[j] for j in NEIGHBORS[vertex]})


def choose_vertex(state, previous=40):
    """DSATUR: highest neighbor-color saturation; spatial ties stabilize the display."""
    uncolored = [i for i, value in enumerate(state) if value == 0]
    if not uncolored:
        return None

    def key(i):
        saturation = len({state[j] for j in NEIGHBORS[i] if state[j]})
        distance = (i // 9 - previous // 9) ** 2 + (i % 9 - previous % 9) ** 2
        return saturation, -distance, -i

    return max(uncolored, key=key)


def solve(puzzle=DEMO, *, check_unique=True, event_limit=100000, timeout=20.0):
    """Find an exact 9-coloring, retaining every assignment, dead end, and undo.

    No answer grid is accepted or stored. Colors are chosen solely from adjacent
    vertices. The uniqueness search continues after the first solution without
    changing the saved first-solution trace. A time limit is reported explicitly.
    """
    puzzle = parse_puzzle(puzzle)
    validate_givens(puzzle)
    state = puzzle.copy()
    events = []
    answers = []
    first_trace = None
    limit = 2 if check_unique else 1
    start = time.perf_counter()
    stats = {"search_nodes": 0, "assignments": 0, "undos": 0, "dead_ends": 0}

    def event(kind, vertex, color, previous_color, candidates, before):
        if first_trace is not None:
            return
        if len(events) >= event_limit:
            raise TimeoutError(
                f"Trace exceeded {event_limit} events; no result has been fabricated."
            )
        events.append(
            {
                "kind": kind,
                "vertex": vertex,
                "color": color,
                "previous_color": previous_color,
                "allowed": list(candidates),
                "before": before,
                "after": state.copy(),
            }
        )

    def visit(previous):
        nonlocal first_trace
        stats["search_nodes"] += 1
        if time.perf_counter() - start > timeout:
            raise TimeoutError(f"Search exceeded {timeout:g} seconds.")
        vertex = choose_vertex(state, previous)
        if vertex is None:
            answers.append(state.copy())
            if first_trace is None:
                first_trace = list(events)
            return len(answers) >= limit
        choices = legal_colors(state, vertex)
        if not choices:
            stats["dead_ends"] += 1
            event("dead_end", vertex, 0, 0, [], state.copy())
            return False
        for color in choices:
            before = state.copy()
            state[vertex] = color
            stats["assignments"] += 1
            event("assign", vertex, color, 0, choices, before)
            if visit(vertex):
                return True
            before = state.copy()
            state[vertex] = 0
            stats["undos"] += 1
            event("undo", vertex, 0, color, choices, before)
        return False

    exhausted = False
    search_error = None
    try:
        exhausted = not visit(40)
    except TimeoutError as exc:
        search_error = str(exc)

    trace = first_trace if first_trace is not None else events
    solution = answers[0] if answers else None
    status = "solved" if solution else "incomplete" if search_error else "unsatisfiable"
    if check_unique and exhausted and not search_error:
        unique = len(answers) == 1
    elif len(answers) >= 2:
        unique = False
    else:
        unique = None
    result = {
        "schema": 1,
        "status": status,
        "algorithm": "DSATUR exact 9-coloring with backtracking",
        "puzzle": puzzle,
        "solution": solution,
        "trace": trace,
        "unique": unique,
        "solutions_found": len(answers),
        "search_exhausted": exhausted,
        "solution_count_limit": limit,
        "error": search_error,
        "graph": {
            "vertices": 81,
            "edges": EDGES,
            "edge_count": len(EDGES),
            "degrees": [len(x) for x in NEIGHBORS],
            "colors": 9,
        },
        "stats": {
            **stats,
            "elapsed_ms": round((time.perf_counter() - start) * 1000, 3),
            "recorded_assignments": sum(e["kind"] == "assign" for e in trace),
            "recorded_undos": sum(e["kind"] == "undo" for e in trace),
            "recorded_dead_ends": sum(e["kind"] == "dead_end" for e in trace),
        },
        "solver_sha256": hashlib.sha256(Path(__file__).read_bytes()).hexdigest(),
    }
    result["trace_sha256"] = hashlib.sha256(
        json.dumps(trace, separators=(",", ":")).encode()
    ).hexdigest()
    return result


def main():
    p = argparse.ArgumentParser(description=__doc__)
    source = p.add_mutually_exclusive_group()
    source.add_argument("--puzzle", help="81 digits, with 0 or . for blanks")
    source.add_argument("--puzzle-file", type=Path)
    p.add_argument(
        "--out", type=Path, help="Write the full graph, solution and actual trace as JSON"
    )
    p.add_argument("--first-only", action="store_true", help="Skip the separate uniqueness search")
    p.add_argument("--timeout", type=float, default=20.0)
    args = p.parse_args()
    puzzle = args.puzzle_file.read_text() if args.puzzle_file else args.puzzle or DEMO
    try:
        result = solve(puzzle, check_unique=not args.first_only, timeout=args.timeout)
    except PuzzleError as exc:
        p.error(str(exc))
    if args.out:
        args.out.parent.mkdir(parents=True, exist_ok=True)
        args.out.write_text(json.dumps(result, indent=2) + "\n")
    print(f"{result['status']}; vertices=81 edges=810; unique={result['unique']}")
    if result["solution"]:
        for r in range(9):
            print(" ".join(map(str, result["solution"][r * 9 : r * 9 + 9])))
    print(json.dumps(result["stats"], sort_keys=True))
    if result["error"]:
        print(result["error"])
    if args.out:
        print(args.out.resolve())
    raise SystemExit(0 if result["status"] == "solved" else 2)


if __name__ == "__main__":
    main()

File: web/solver.js
/* Exact Sudoku graph coloring. No libraries, answer tables, or remote calls. */
'use strict';

function createSudokuSolver() {
  const peers = Array.from({length: 81}, () => []);
  const edges = [];
  const box = i => Math.floor(i / 27) * 3 + Math.floor(i % 9 / 3);
  for (let i = 0; i < 81; i++) {
    for (let j = i + 1; j < 81; j++) {
      if (Math.floor(i / 9) === Math.floor(j / 9) || i % 9 === j % 9 || box(i) === box(j)) {
        peers[i].push(j);
        peers[j].push(i);
        edges.push([i, j]);
      }
    }
  }

  function parsePuzzle(value) {
    if (typeof value === 'string') {
      const text = value.replace(/[\s|+\-]/g, '');
      if (!/^[0-9.]{81}$/.test(text)) {
        throw new Error('Enter exactly 81 digits; use 0 or . for an empty cell.');
      }
      return Array.from(text, c => c === '.' ? 0 : Number(c));
    }
    if (!Array.isArray(value) || value.length !== 81 ||
        Array.from(value).some(v => !Number.isInteger(v) || v < 0 || v > 9)) {
      throw new Error('A puzzle must contain 81 integers from 0 to 9.');
    }
    return value.slice();
  }

  function validateGivens(puzzle) {
    for (const [i, j] of edges) {
      if (puzzle[i] && puzzle[i] === puzzle[j]) {
        throw new Error(`Conflicting givens at r${Math.floor(i / 9) + 1}c${i % 9 + 1} and ` +
          `r${Math.floor(j / 9) + 1}c${j % 9 + 1}: both are ${puzzle[i]}.`);
      }
    }
  }

  function solve(input, options = {}, onProgress = () => {}) {
    const puzzle = parsePuzzle(input);
    validateGivens(puzzle);
    const checkUnique = options.checkUnique !== false;
    const timeoutMs = options.timeoutMs ?? 30000;
    const maxTraceEvents = options.maxTraceEvents ?? 20000;
    if (!Number.isFinite(timeoutMs) || timeoutMs < 0 ||
        !Number.isInteger(maxTraceEvents) || maxTraceEvents < 0) {
      throw new Error('Timeout and trace limit must be nonnegative; the trace limit must be an integer.');
    }
    const state = puzzle.slice();
    const trace = [], answers = [];
    const stats = {search_nodes: 0, assignments: 0, undos: 0, dead_ends: 0};
    const started = performance.now();
    let lastProgress = started, firstFound = false, truncated = false, searchError = null;
    const limit = checkUnique ? 2 : 1;

    function record(kind, vertex, color, previousColor, choices, before) {
      if (firstFound) return;
      if (trace.length >= maxTraceEvents) {
        truncated = true;
        return;
      }
      trace.push({kind, vertex, color, previous_color: previousColor,
        allowed: choices.slice(), before, after: state.slice()});
    }

    function visit(previous) {
      stats.search_nodes++;
      const now = performance.now();
      if (timeoutMs && now - started >= timeoutMs) {
        throw new Error(`Search reached the ${timeoutMs / 1000} second limit.`);
      }
      if (now - lastProgress >= 150) {
        lastProgress = now;
        onProgress({...stats, elapsed_ms: now - started, solutions_found: answers.length});
      }

      let vertex = -1, bestSaturation = -1, bestDistance = Infinity, chosenMask = 0;
      for (let i = 0; i < 81; i++) {
        if (state[i]) continue;
        let used = 0;
        for (const j of peers[i]) if (state[j]) used |= 1 << state[j];
        let bits = used, saturation = 0;
        while (bits) { bits &= bits - 1; saturation++; }
        const distance = (Math.floor(i / 9) - Math.floor(previous / 9)) ** 2 +
          (i % 9 - previous % 9) ** 2;
        if (saturation > bestSaturation || (saturation === bestSaturation && distance < bestDistance)) {
          vertex = i;
          bestSaturation = saturation;
          bestDistance = distance;
          chosenMask = used;
        }
      }
      if (vertex === -1) {
        answers.push(state.slice());
        firstFound = true;
        return answers.length >= limit;
      }
      const choices = [];
      for (let color = 1; color <= 9; color++) if (!(chosenMask & (1 << color))) choices.push(color);
      if (!choices.length) {
        stats.dead_ends++;
        record('dead_end', vertex, 0, 0, [], state.slice());
        return false;
      }
      for (const color of choices) {
        const before = !firstFound && !truncated ? state.slice() : null;
        state[vertex] = color;
        stats.assignments++;
        record('assign', vertex, color, 0, choices, before);
        if (visit(vertex)) return true;
        const assigned = !firstFound && !truncated ? state.slice() : null;
        state[vertex] = 0;
        stats.undos++;
        record('undo', vertex, 0, color, choices, assigned);
      }
      return false;
    }

    let exhausted = false;
    try { exhausted = !visit(40); }
    catch (error) {
      if (!error.message.startsWith('Search reached the ')) throw error;
      searchError = error.message;
    }
    const solution = answers[0] || null;
    const unique = answers.length >= 2 ? false : checkUnique && exhausted && !searchError ? answers.length === 1 : null;
    return {
      schema: 1, implementation: 'JavaScript', status: solution ? 'solved' : searchError ? 'incomplete' : 'unsatisfiable',
      algorithm: 'DSATUR exact 9-coloring with backtracking', puzzle, solution, trace,
      trace_complete: !truncated, unique, solutions_found: answers.length,
      search_exhausted: exhausted, solution_count_limit: limit, error: searchError,
      graph: {vertices: 81, edges, edge_count: edges.length, degrees: peers.map(p => p.length), colors: 9},
      stats: {...stats, elapsed_ms: Math.round((performance.now() - started) * 1000) / 1000,
        recorded_assignments: trace.filter(e => e.kind === 'assign').length,
        recorded_undos: trace.filter(e => e.kind === 'undo').length,
        recorded_dead_ends: trace.filter(e => e.kind === 'dead_end').length}
    };
  }
  return {solve, parsePuzzle, validateGivens, peers, edges};
}

if (typeof module !== 'undefined' && module.exports) module.exports = {createSudokuSolver};
